### ISSUE TITLE
[LifetimeSafety] False negative for `a = T(s)` when T's constructor has [[clang::lifetimebound]]

### ISSUE THREAD

**Issue by aeft** [llvm/llvm-project#190912]
Following #187917, the analysis tracks origins for record types returned from `[[clang::lifetimebound]]` calls. However, when such a construction appears as a functional-cast expression (e.g. `StringView(s)` used as an rvalue), the loan is dropped and the resulting dangle is silently missed.

https://godbolt.org/z/ofT9cef8h
```cpp
#include <string>

struct StringView {
  StringView();
  StringView(const std::string &s [[clang::lifetimebound]]);
};

void case_a() {
  StringView a;
  {
    std::string s = "temp";
    StringView tmp(s);
    a = tmp; // warn correctly
  }
  (void)a;
}

void case_b() {
  StringView a;
  {
    std::string s = "temp";
    a = StringView(s); // should warn
  }
  (void)a;
}
```

AST
```
    |     |-DeclRefExpr 0x1580cdd90 <col:5> 'StringView' lvalue Var 0x1580cdb40 'a' 'StringView'
    |     `-MaterializeTemporaryExpr 0x1580cdf40 <col:9, col:21> 'StringView' xvalue
    |       `-CXXFunctionalCastExpr 0x1580cdf18 <col:9, col:21> 'StringView' functional cast to StringView <ConstructorConversion>
    |         `-CXXConstructExpr 0x1580cddf8 <col:9, col:21> 'StringView' 'void (const std::string &)'
    |           `-ImplicitCastExpr 0x1580cdde0 <col:20> 'const std::string' lvalue <NoOp>
    |             `-DeclRefExpr 0x1580cddc0 <col:20> 'std::string' lvalue Var 0x1580cdc30 's' 'std::string'
```

Currently, `VisitCXXFunctionalCastExpr` only forwards origins for `gsl::Pointer` types.
